File: ⑤二次开发例程/⑤二次开发例程/C# 例子_x64/C# 2010例子_x64/ECanTest/ECanTest/ComProc.cs
using System;
using System.Collections.Generic;
using System.Text;
using ECAN;
using System.Threading;

namespace ECanTest
{
    class ComProc
    {


        // Fields
        public bool EnableProc;

        public const int REC_MSG_BUF_MAX = 0x2710;

        public CAN_OBJ[] gRecMsgBuf;
        public uint gRecMsgBufHead;
        public uint gRecMsgBufTail;


        public const int SEND_MSG_BUF_MAX = 0x2710;

        public CAN_OBJ[] gSendMsgBuf;
        public uint gSendMsgBufHead;
        public uint gSendMsgBufTail;



        private Timer _RecTimer;
        private Timer _SendTimer;

        private AutoResetEvent RecEvent;
        private TimerCallback RecTimerDelegate;
        private AutoResetEvent SendEvent;
        private TimerCallback SendTimerDelegate;


      
        public ComProc()
        {
            this.gSendMsgBuf = new CAN_OBJ[SEND_MSG_BUF_MAX];
            this.gSendMsgBufHead = 0;
            this.gSendMsgBufTail = 0;


            this.gRecMsgBuf = new CAN_OBJ[REC_MSG_BUF_MAX];
            this.gRecMsgBufHead = 0;
            this.gRecMsgBufTail = 0;



            this.EnableProc = false;
            this.RecEvent = new AutoResetEvent(false);
            this.RecTimerDelegate = new TimerCallback(this.RecTimer_Tick);
            this._RecTimer = new Timer(this.RecTimerDelegate, this.RecEvent, 0, 20);
            this.SendEvent = new AutoResetEvent(false);
            this.SendTimerDelegate = new TimerCallback(this.SendTimer_Tick);
            this._SendTimer = new Timer(this.SendTimerDelegate, this.SendEvent, 0, 20);

        }


 

        private void ReadMessages()
        {
            CAN_OBJ mMsg = new CAN_OBJ();

            int sCount = 0;
            do
            {
                uint mLen = 1;
                if (!(ECANDLL.Receive(1, 0, 0, out mMsg, mLen, 1) == ECANStatus.STATUS_OK))
                {
                    break;
                }
                if (mLen == 0) break;

                this.gRecMsgBuf[this.gRecMsgBufHead].ID = mMsg.ID;
                this.gRecMsgBuf[this.gRecMsgBufHead].DataLen = mMsg.DataLen;
                this.gRecMsgBuf[this.gRecMsgBufHead].data = mMsg.data;
                this.gRecMsgBuf[this.gRecMsgBufHead].ExternFlag = mMsg.ExternFlag;
                this.gRecMsgBuf[this.gRecMsgBufHead].RemoteFlag = mMsg.RemoteFlag;
				this.gRecMsgBuf[this.gRecMsgBufHead].TimeStamp = mMsg.TimeStamp;
                this.gRecMsgBuf[this.gRecMsgBufHead].Reserved = mMsg.Reserved;
                this.gRecMsgBuf[this.gRecMsgBufHead].TimeFlag = mMsg.TimeFlag;
                this.gRecMsgBufHead += 1;
                if (this.gRecMsgBufHead >= REC_MSG_BUF_MAX)
                {
                    this.gRecMsgBufHead = 0;
                }
                sCount++;
            }
            while (sCount < 500);
        }



 
        private void SendMessages()
        {
            CAN_OBJ[] mMsg = new CAN_OBJ[2];
            int aa=0;

            int sCount = 0;
            do
            {
                if (this.gSendMsgBufHead == this.gSendMsgBufTail)
                {
                    break;
                }
                mMsg[0] = this.gSendMsgBuf[this.gSendMsgBufTail];
                mMsg[1] = this.gSendMsgBuf[this.gSendMsgBufTail];
                //mMsg[0].data[0] = 0;
                //mMsg[0].data[1] = 0;
                //mMsg[0].data[2] = 0;
                this.gSendMsgBufTail+=1;
                if (this.gSendMsgBufTail >= SEND_MSG_BUF_MAX)
                {
                    this.gSendMsgBufTail = 0;
                }
                uint mLen = 1;
                aa=System.Runtime.InteropServices.Marshal.SizeOf(mMsg[1]);
                if (ECANDLL.Transmit(1, 0, 0, mMsg, (ushort)mLen) != ECANStatus.STATUS_OK)
                {
                }
                sCount++;
            }
            while (sCount < 200);
        }

      
        public void RecTime()
        {
            if (this.EnableProc)
            {
                this.ReadMessages();
            }
        }

        private void RecTimer_Tick(object mObject)
        {
            this.RecTime();
        }

        private void SendTimer_Tick(object mObject)
        {
            this.SendTime();
        }


        public void SendTime()
        {
            if (this.EnableProc)
            {
                this.SendMessages();
            }
        }




        public void Close()
        {
        }

 
    }
}

File: ⑤二次开发例程/⑤二次开发例程/VBNET例子/ECanTest2005_Thread/Backup/ECanTest/ComProc.vb
Imports System.Threading

Public Class ComProc

    Public Const SEND_MSG_BUF_MAX = 10000
    Public gSendMsgBuf(SEND_MSG_BUF_MAX) As ECAN.CAN_OBJ
    Public gSendMsgBufHead As UInt32 = 0
    Public gSendMsgBufTail As UInt32 = 0


    Public Const REC_MSG_BUF_MAX = 10000
    Public gRecMsgBuf(REC_MSG_BUF_MAX) As ECAN.CAN_OBJ
    Public gRecMsgBufHead As UInt32 = 0
    Public gRecMsgBufTail As UInt32 = 0



    Private RecCount_1 As UInt16 = 0
    Private recFirst_1 As Int16 = 1


    Public EnableProc As Boolean = False


    Private _RecTimer As Threading.Timer
    Private RecEvent As Threading.AutoResetEvent
    Private RecTimerDelegate As Threading.TimerCallback


    Private _SendTimer As Threading.Timer
    Private SendEvent As Threading.AutoResetEvent
    Private SendTimerDelegate As Threading.TimerCallback


    Sub New()


        RecEvent = New Threading.AutoResetEvent(False)
        RecTimerDelegate = AddressOf RecTimer_Tick
        _RecTimer = New Threading.Timer(RecTimerDelegate, RecEvent, 0, 20)


        SendEvent = New Threading.AutoResetEvent(False)
        SendTimerDelegate = AddressOf SendTimer_Tick
        _SendTimer = New Threading.Timer(SendTimerDelegate, SendEvent, 0, 20)


    End Sub


    Private Sub RecTimer_Tick(ByVal mObject As Object)

        RecTime()

    End Sub

    Private Sub SendTimer_Tick(ByVal mObject As Object)
        SendTime()
    End Sub


    Public Sub RecTime()
        If EnableProc Then
            ReadMessages()

        End If
    End Sub

    Public Sub SendTime()
        If EnableProc Then
            SendMessages()
        End If

    End Sub

    Private Sub ReadMessages()



        Dim stsResult As ECan.ECANStatus
        Dim mMsg As ECAN.CAN_OBJ = New ECAN.CAN_OBJ
        Dim mLen As UInt32
        Dim overlp As Overlapped = New Overlapped
        Dim sCount As Integer = 0

        '			
        Do
            '
            mLen = 1
            stsResult = ECAN.ECANDLL.Receive(1, 0, 0, mMsg, mLen, 20)

            '
            If stsResult = ECAN.ECANStatus.STATUS_OK And mLen > 0 Then

                Dim recMSG As ECAN.CAN_OBJ = New ECAN.CAN_OBJ
                gRecMsgBuf(gRecMsgBufHead).ID = mMsg.ID
                gRecMsgBuf(gRecMsgBufHead).DataLen = mMsg.DataLen
                gRecMsgBuf(gRecMsgBufHead).data = mMsg.data
                gRecMsgBuf(gRecMsgBufHead).ExternFlag = mMsg.ExternFlag
                gRecMsgBuf(gRecMsgBufHead).RemoteFlag = mMsg.RemoteFlag

                gRecMsgBufHead = gRecMsgBufHead + 1
                If (gRecMsgBufHead) >= REC_MSG_BUF_MAX Then
                    gRecMsgBufHead = 0
                End If



            Else

                Exit Do

            End If

            sCount = sCount + 1
            If sCount >= 500 Then
                Exit Do

            End If

        Loop


    End Sub

  

    Private Sub SendMessages()
        Dim stsResult As ECan.ECANStatus
        Dim mMsg As ECAN.CAN_OBJ
        Dim mLen As UInt32
        Dim sCount As Integer = 0

        Do
            '


            If gSendMsgBufHead = gSendMsgBufTail Then

                Exit Do

            End If

            mMsg = gSendMsgBuf(gSendMsgBufTail)
            gSendMsgBufTail = gSendMsgBufTail + 1
            If gSendMsgBufTail >= SEND_MSG_BUF_MAX Then
                gSendMsgBufTail = 0
            End If

            mLen = 1
            stsResult = ECAN.ECANDLL.Transmit(1, 0, 0, mMsg, mLen)


            If stsResult <> ECAN.ECANStatus.STATUS_OK Then

                '             msgbox("Transmit send fault" )

            End If
            sCount = sCount + 1


            If sCount >= 200 Then
                Exit Do

            End If


        Loop
    End Sub

    
    Public Sub Close()


    End Sub
End Class
